Pocket/pocket-ios
# Improve scrolling performance of Archive

## Summary
This PR improves the scrolling performance of archived content. When fetching data from the server, we can only fetch ~30 items at a time, which means we have to use pagination. But we want users to be able to scroll continuously through their list, keeping the implementation detail of pagination transparent to the user.

## References 
https://getpocket.atlassian.net/browse/IN-533

## Implementation Details
We take advantage of the `totalCount` of archived items that is provided by the server each time we fetch a page. We can build a collection view with a cell for each item that has been fetched and stored locally, and then create "placeholder" cells for the remaining items that have not yet been fetched.

Then we can take advantage of [UIKit's api for prefetching collection view data](https://developer.apple.com/documentation/uikit/uicollectionviewdatasourceprefetching/prefetching_collection_view_data) that will allow us to fetch data on demand. The logic for managing these prefetching requests is encapsulated in `Sync.ArchiveService`.

The ArchiveService is responsible for taking in these prefetch requests, queuing the requests in an operation queue, and storing the results locally. It then publishes an array of `ArchivedItem` objects that can be consumed by the view model, and ultimately published to the view controller for rendering.

(see commit message for more a list of changes in this commit)

## Test Steps

1. Navigate to My List tab, tap "Archive" in the My List/Archive switcher2. 
2. Scroll down quickly
3. Content should load in as needed and allow a smooth scrolling experience.
4. Verify that the "Favorites" and "All" buttons work as expected.



## TODO
There are a few items left to tidy up before this is fully ready to be merged:

- [ ] Fixup methods for tracking impressions and item actions in ArchivedItemsViewModel
- [ ] Add “skeleton” for pending cells
- [ ] Add methods to `ItemsListViewModel` to allow calling data prefetching methods from view controller without casting to concrete type `ArchivedItemsListViewModel`
